### ISSUE TITLE
S3 Endpoint Mapping

### ISSUE THREAD

**Issue by saocean** [cherweg/logstash-input-s3-sns-sqs#15]
We have a need to specify the S3 endpoint in our configuration. I see this can be added but I believe it's only applicable to the SQS portion of the native logstash-input-sqs plugin.

Is there a way we can expose the endpoint parameter for specifying a different s3 endpoint?

For example:

endpoint => "https://s3-fips.us-east-2.amazonaws.com"

**Comment by christianherweg0807** [comment]
You´d like to overwrite the s3 Endpoint?  Normally this the bucket arn is included in the SNS notification:
`{
   "Type":"Notification",
   "MessageId":"f42cbc8a-1950-54cb-ac64-e9c1909134ab",
   "TopicArn":"arn:aws:sns:eu-central-1:xxx:xxx",
   "Subject":"Amazon S3 Notification",
   "Message":"{"   Records":[
      {
         "s3":{
            "s3SchemaVersion":"1.0",
            "configurationId":"ad3eedd5-3e21-448a-932a-72ad1b859178",
            "bucket":{
               "name":"mybuckename",
               "ownerIdentity":{
                  "principalId":"A15O91HAOOXKYF"
               },
               "arn":"arn:aws:s3:::mybucketname"
            },
            "object":{
               "key":"Folder/file.xxx",
            }
         }
      }
   ]
}`
S3 notifies SNS on file creation and sends the endpoint inside the message.
I can´t identify your usecase... please explain.

regards
christian

**Comment by saocean** [comment]
Hey Christian,

Thanks for taking the time. So this is what I am referring to: https://docs.aws.amazon.com/general/latest/gr/rande.html

In the logstash-s3-input the parameter is "endpoint"

The endpoint to connect to. By default it is constructed using the value of region. This is useful when connecting to S3 compatible services, but beware that these aren’t guaranteed to work correctly with the AWS SDK.

We've used it with that plugin without issues. Essentially it changes the default endpoint which I  is constructed from the service and region (https://s3.us-east-2.amazonaws.com) to something I specify (https://s3-fips.us-east-2.amazonaws.com)

**Comment by christianherweg0807** [comment]
Hi this enpoint mapping is an logstash mixin feature, i partialy use for connecting to aws services inside this plugin. 
https://github.com/logstash-plugins/logstash-mixin-aws/blob/7cb4b6092221c4a9156e194c6cd6032788bb9206/lib/logstash/plugin_mixins/aws_config/generic.rb#L29

Maybe it is a very small change, when establishing the s3 connection here:
https://github.com/cherweg/logstash-input-s3-sns-sqs/blob/41a06c35d922b7d2bc56cdd563af70fead9e54d8/lib/logstash/inputs/s3snssqs.rb#L397-L406

I´ve to test this locally...

regards
Christian

**Comment by saocean** [comment]
Thanks man, let us know if you need anything from our end.

**Comment by christianherweg0807** [comment]
Going a little bit deeper into to logstash::mixin code, i think simply using the endpoint option should work. Have you ever tried to do?

**Comment by saocean** [comment]
Hey Christian,

I think that should work fine for the s3 endpoint, doing something like s3_endpoint. I know if we specified "endpoint" as an option, it didn't work with your plugin when we tested. I think because the plugin is using it for the sqs endpoint instead of the s3 endpoint because the "endpoint" option is calling the underlying logstash-sqs-input.

As an example, the reason I think it's using "endpoint" only for sqs, when we needed to assume a role cross account for sqs and s3 with your plugin, we actually did something similar and specified the role twice. Once for s3 using the "s3_role_arn" and once sqs using "role_arn". role_arn was used by the native sqs plugin for the assume role of sqs and the s3_role_arn was used for the s3 portion. We could not get it to work without specifying the role twice using those two options.

This setup actually worked great for us because sometimes we place our sqs queue in one account and while the s3 bucket is in another account. This allowed use to use 2 separate roles in two separate accounts.

Anyway using that example of how we need to have a specific portion for assume role called out in the s3 and sqs portion, is why i think you would need a specific portion for endpoint in the s3 (doesn't exist yet) and sqs (already exists for sqs) portions.

Sorry that got long and overly complicated. I was trying to provide examples to my thinking. We definitely appreciate all the help.

**Comment by christianherweg0807** [comment]
Ahh...now I understand. We´ve to do something like a per aws-service options-hash. 
```
aws_options_hash_by_Service: => 
 {"Aws::SQS::Client" => 
 "{endpoint => "https://s3-fips.us-east-2.amazonaws.com"}"} 
```

**Comment by saocean** [comment]
That sounds good to me! Let me know if you need anything from us.

**Comment by saocean** [comment]
Hey Christian,

I saw you just put a bunch of work you just put in. Looks awesome!

Was the ability to specify the s3 endpoint added as part of the "s3_options_by_bucket"? Would you be able to provide an example just for my clarification using s3_role_arn, bucket, folder etc. I see some of an example in s3snssqs.rb

**Comment by cherweg** [comment]
Hello,
I forgot a commit, to solve this issue. 
We implemented an option to add s3 default options in your logstash config:
s3_default_options          => { "endpoint" => "https://account-id....." }

With these option you could merge every option in the aws_options_hash for the s3 client. 
See: https://docs.aws.amazon.com/sdkforruby/api/Aws/S3/Client.html

Plz test & use with care...


regards Christian

**Comment by cherweg** [comment]
https://github.com/cherweg/logstash-input-s3-sns-sqs/releases/tag/1.9.9

**Comment by cherweg** [comment]
I´ll release this soon at rubygems.org for direkt installation in logstash...

**Comment by christianherweg0807** [comment]
** Execute publish_gem
Published logstash-input-s3-sns-sqs-2.0.0.gem
